### ISSUE TITLE
JAX scan accumulated output slows down the loops

### ISSUE THREAD

**Issue by AdrienCorenflos** [google/jax#4811]
Hi,

I have noticed that the computational graph may be wrongly traced for scan operations where the accumulated output is not returned, for example, if I take:

```python
def sqrt_not_careful(x):
    def body(val, _):
        new_val = 0.5 * (val + x / val)
        return new_val, new_val
    
    res, _ = scan(body, jnp.ones_like(x), jnp.arange(10))
    return res

def sqrt_careful(x):
    def body(val, _):
        new_val = 0.5 * (val + x / val)
        return new_val, None
    
    res, _ = scan(body, jnp.ones_like(x), jnp.arange(10))
    return res
```
then `sqrt_not_careful` can be orders of magnitudes slower than `sqrt_careful`. My original assumption would have been that once jitted, the two behaviours should have been the same, but maybe I'm missing the technical background to understand why it's not the case.

**Comment by jakevdp** [comment]
I believe `body` should take two arguments here. Can you double-check the code?

**Comment by AdrienCorenflos** [comment]
Correct, sorry about this I adapted back of the envelope a more complex example. I edited the question so that it would run indeed, and actually doing this, it seems that my remark only applied to non-jitted functions, so I'm closing the issue. 
Apologies for the loss of time.
